src/main/java/com/dataartisans/flink/dataflow/translation/wrappers/streaming/FlinkGroupAlsoByWindowWrapper.java

```diff
@@ -375,114 +378,136 @@ private void unregisterActiveTimer(K key, TimerInternals.TimerData timer) {
-			unregisterActiveTimer(key, timerKey);
+		public void deleteTimer(TimerData timerKey) {
+			unregisterActiveTimer(context.element().key(), timerKey);
 		}
 	}
 
-	private class ProcessContext extends DoFn<TimerOrElement<WindowedValue<KV<K, VIN>>>, KV<K, VOUT>>.ProcessContext {
+	private class ProcessContext extends GroupAlsoByWindowViaWindowSetDoFn<K, VIN, VOUT, ?, KeyedWorkItem<K, VIN>>.ProcessContext {
 
 		private final FlinkTimerInternals timerInternals;
 
-		private final DoFn<TimerOrElement<WindowedValue<KV<K, VIN>>>, KV<K, VOUT>> fn;
-
-		private final Collector<WindowedValue<KV<K, VOUT>>> collector;
+		private final TimestampedCollector<WindowedValue<KV<K, VOUT>>> collector;
 
 		private FlinkStateInternals<K> stateInternals;
 
-		private TimerOrElement<WindowedValue<KV<K, VIN>>> element;
+		private KeyedWorkItem<K, VIN> element;
 
-		public ProcessContext(DoFn<TimerOrElement<WindowedValue<KV<K, VIN>>>, KV<K, VOUT>> function,
-							  Collector<WindowedValue<KV<K, VOUT>>> outCollector,
-							  FlinkTimerInternals timerInternals) {
+		public ProcessContext(DoFn<KeyedWorkItem<K, VIN>, KV<K, VOUT>> function,
+		                      TimestampedCollector<WindowedValue<KV<K, VOUT>>> outCollector,
+		                      FlinkTimerInternals timerInternals) {
 			function.super();
 			super.setupDelegateAggregators();
 
-			this.fn = Preconditions.checkNotNull(function);
 			this.collector = Preconditions.checkNotNull(outCollector);
 			this.timerInternals = Preconditions.checkNotNull(timerInternals);
 		}
 
-		public void setElement(TimerOrElement<WindowedValue<KV<K, VIN>>> value,
-							   FlinkStateInternals<K> stateForKey) {
-			this.element = value;
+		public void setElement(KeyedWorkItem<K, VIN> element,
+		                       FlinkStateInternals<K> stateForKey) {
+			this.element = element;
 			this.stateInternals = stateForKey;
-			this.timerInternals.setElement(value);
 		}
 
-		public void setCurrentWatermark(Instant watermark) {
-			this.timerInternals.setCurrentWatermark(watermark);
+		public void setCurrentInputWatermark(Instant watermark) {
+			this.timerInternals.setCurrentInputWatermark(watermark);
+		}
+
+		public void setCurrentOutputWatermark(Instant watermark) {
+			this.timerInternals.setCurrentOutputWatermark(watermark);
 		}
 
 		@Override
-		public TimerOrElement element() {
-			if (element != null && !this.element.isTimer()) {
-				return TimerOrElement.element(this.element.element().getValue());
-			}
+		public KeyedWorkItem<K, VIN> element() {
 			return this.element;
 		}
 
 		@Override
 		public Instant timestamp() {
-			return this.element.isTimer() ?
-					this.element.getTimer().getTimestamp() :
-					this.element.element().getTimestamp();
+			throw new UnsupportedOperationException("timestamp() is not available when processing KeyedWorkItems.");
 		}
 
 		@Override
 		public PipelineOptions getPipelineOptions() {
+			// TODO: PipelineOptions need to be available on the workers.
```

Comment:
We could pass the PipelineOptions as part of the operator state. We can fix that.